util/oci/client.go

```diff
@@ -0,0 +1,577 @@
+	c := &nativeOCIClient{
+		creds:             creds,
+		repoURL:           repoURL,
+		repoLock:          repoLock,
+		repo:              repo,
+		tagsFunc:          tagsFunc,
+		allowedMediaTypes: layerMediaTypes,
+	}
+	for i := range opts {
+		opts[i](c)
+	}
+	return c
+}
+
+// nativeOCIClient implements Client interface using oras-go
+type nativeOCIClient struct {
+	creds                           Creds
+	repoURL                         string
+	repo                            oras.ReadOnlyTarget
+	tagsFunc                        func(context.Context, string) ([]string, error)
+	repoLock                        sync.KeyLock
+	tagsCache                       tagsCache
+	repoCachePaths                  argoio.TempPaths
+	allowedMediaTypes               []string
+	manifestMaxExtractedSize        int64
+	disableManifestMaxExtractedSize bool
+}
+
+// TestRepo verifies that the remote OCI repo can be connected to.
+func (c *nativeOCIClient) TestRepo(ctx context.Context) (bool, error) {
+	_, err := c.tagsFunc(ctx, "")
+	return err == nil, err
+}
+
+func (c *nativeOCIClient) Extract(ctx context.Context, digest string, project string) (string, argoio.Closer, error) {
+	cachedPath, err := c.getCachedPath(digest, project)
+	if err != nil {
+		return "", nil, fmt.Errorf("error getting oci path for digest %s: %w", digest, err)
+	}
+
+	c.repoLock.Lock(cachedPath)
+	defer c.repoLock.Unlock(cachedPath)
+
+	exists, err := fileExists(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if !exists {
+		ociManifest, err := getOCIManifest(ctx, digest, c.repo)
+		if err != nil {
+			return "", nil, err
+		}
+
+		if len(ociManifest.Layers) != 1 {
+			return "", nil, fmt.Errorf("expected only a single oci layer, got %d", len(ociManifest.Layers))
+		}
+
+		if !slices.Contains(c.allowedMediaTypes, ociManifest.Layers[0].MediaType) {
+			return "", nil, fmt.Errorf("oci layer media type %s is not in the list of allowed media types", ociManifest.Layers[0].MediaType)
+		}
+
+		err = saveCompressedImageToPath(ctx, digest, c.repo, cachedPath)
+		if err != nil {
+			return "", nil, fmt.Errorf("could not save oci digest %s: %w", digest, err)
+		}
+	}
+
+	maxSize := c.manifestMaxExtractedSize
+	if c.disableManifestMaxExtractedSize {
+		maxSize = math.MaxInt64
+	}
+
+	manifestsDir, err := extractContentToManifestsDir(ctx, cachedPath, digest, maxSize, c.allowedMediaTypes)
+	if err != nil {
+		return manifestsDir, nil, fmt.Errorf("cannot extract contents of oci image with revision %s: %w", digest, err)
+	}
+
+	return manifestsDir, argoio.NewCloser(func() error {
+		return os.RemoveAll(manifestsDir)
+	}), nil
+}
+
+func (c *nativeOCIClient) getCachedPath(version, project string) (string, error) {
+	keyData, err := json.Marshal(map[string]string{"url": c.repoURL, "project": project, "version": version})
```

Comment:
Why key on the version? Doesn't git/helm just key on repo url and project?

Reply:
I copied what Helm did, which also [keys](https://github.com/argoproj/argo-cd/blob/d016dbec4d74a888a4564fd30ed8892b7476ba36/util/helm/client.go#L391) on repo url, revision and project. 

As with Helm, the cached path has a reference to a tarball on disk that contains the actual manifests.

Reply:
This was fixed to align with #22712